Allow 'cylc reload' to reload global configuration

ScottWales opened the pull request:
<!--

Thanks for your contribution! Please:
* List any related issues with a "closes" or "addresses" tag.
* Add a helpful title & description.
* Complete the checklist.

-->

Add a `--global` flag to `cylc reload` that will reload the server's global configuration when set. This can be used to update platform definitions while a workflow is running.

Note that:
* Most changes to the `[scheduler]` section (ports, process pool size, etc.) require a server restart to take effect.
* Already installed platforms cannot have their symlink paths changed.
* Some changes to platforms (such as changing the hosts) may prevent Cylc from communicating with existing jobs (through polls, kill etc.).

Closes #3762

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/815
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

oliver-sanders commented:
> Most changes to the [scheduler] section (ports, process pool size, etc.) require a server restart to take effect.

We could do with documenting what can and cannot be reloaded.

The configuration docs are built from the `cfgspec` module which use RestructuredText markup. Could use a `.. note:: Reloading` directive or something of the ilk.

https://github.com/cylc/cylc-flow/blob/9ff50f82ffcb205ab1cb1236ab20ae442a8e6391/cylc/flow/cfgspec/globalcfg.py#L739-L741

---

> Some changes to platforms (such as changing the hosts) may prevent Cylc from communicating with existing jobs (through polls, kill etc.).

That sounds pretty bad! Presumably caused by platform definitions lingering from the original config? Or is it more subtle?

ScottWales commented on cylc/flow/task_proxy.py:
```diff
@@ -344,7 +344,6 @@ def copy_to_reload_successor(
         reload_successor.summary = self.summary
         reload_successor.local_job_file_path = self.local_job_file_path
         reload_successor.try_timers = self.try_timers
-        reload_successor.platform = self.platform
```
If this line is kept the test fails with the reload not updating the platform. Manually testing shows it works with or without this line in the real suite though, it's probable the test isn't checking what I think it should be.

Looking through the source code I'm a bit confused about how platforms get attached to tasks - at [task_proxy.py#L273](https://github.com/cylc/cylc-flow/blob/7f3bd41928f3363465c3982bb3e30b527a4eb28d/cylc/flow/task_proxy.py#L273) no platform name is given to `get_platform`, which as you say should give it the `localhost` platform. I'm not seeing how the proper platform gets attached after that?

oliver-sanders replied:
The platform gets set at submission time:

```console
[cylc] $ git grep '\.platform ='
cylc/flow/run_modes/dummy.py:    itask.platform = get_platform()
cylc/flow/run_modes/simulation.py:    itask.platform = {
cylc/flow/run_modes/skip.py:    itask.platform = {
cylc/flow/task_job_mgr.py:                        itask.platform = platform
cylc/flow/task_job_mgr.py:                itask.platform = platform
cylc/flow/task_pool.py:                    itask.platform = get_platform(platform_name)
cylc/flow/task_proxy.py:            self.platform = {}
cylc/flow/task_proxy.py:            self.platform = get_platform()
cylc/flow/task_proxy.py:        reload_successor.platform = self.platform
cylc/flow/tui/overlay.py:            'fn + ⌥ & click to select text' if sys.platform == 'darwin' else
tests/integration/test_task_job_mgr.py:        schd.pool.get_tasks()[0].platform = {
```

We shouldn't allocate a platform when the `TaskProxy` object is created, this is a misnomer / technical dept yet to be cleaned up.

oliver-sanders commented:
If `a.example.com` and `b.example.com` are both hosts belong to the same platform, we should be able to change the hosts in the way you outline above without error.

If these hosts belong to two different platforms with two discrete queues, then failure is correct, this isn't a valid / safe global config change.

oliver-sanders commented:
How does this functionality play with environment variables defined in the `rose-suite.conf` file?

See https://github.com/cylc/cylc-flow/issues/5418#issuecomment-2304167637

ScottWales commented:
Any environment variables that the cylc-rose plugin sets is kept constant through the reload

ScottWales commented:
It's not clear to me why the functional test is failing. Could be a timing thing, the workflow file is trying to do a cylc pause but that fails as it's using the wrong variable name, should be `CYLC_WORKFLOW_ID` I think.
https://github.com/cylc/cylc-flow/blob/e3f77a5744fbade763842c9bdfb4fdae71c760ba/tests/functional/cylc-combination-scripts/vr_workflow/flow.cylc#L6-L8

oliver-sanders commented:
Yes, that environment variable doesn't look right, but how was the test passing before?

@wxtim, could you take a look at that functional test when you get a chance.

wxtim commented:
> Yes, that environment variable doesn't look right, but how was the test passing before?

1. `run_ok cylc vip ...` was being run in detached mode - so workflow failure won't show up in test one.
2. The test polls for workflow running, but not for tasks succeeding.
3. `run_ok cylc vr ...` runs in a non-interactive shell, so doesn't return 1 because no changes to source have been made. 

In short, this test is garbage.

ScottWales commented:
Thanks Tim, that's fixed the tests

oliver-sanders commented:
Reload has been a notoriously difficult feature to maintain and a [major source of bugs](https://github.com/cylc/cylc-flow/issues?q=is%3Aissue+reload+label%3Abug), so we need to flesh out exactly what does and does not get reloaded and ensure we pin down our desired behavior with tests.

Result of a quick skim of the global config (recommend a more detailed inspection):

* `[scheduler]` - documented as not reloadable, good.
  * `[scheduler][events]` - From code inspection, this will probably get reloaded contrary to the documentation? Needs testing.
  * `[scheduler][mail]` - From code inspection, this will probably get reloaded contrary to the documentation? Needs testing.
* `[install][symlink dirs]` - documented as not reloadable, good.
* `[platforms]` - tested platform metadata gets reloaded in an integration test, but haven't tested that jobs get submitted with the new platform config post-reload. This is an important use case and job submission is fiddly, so we could do with a formal test to ensure this works now and doesn't get broken in the future.
* `[platform groups]` - Not tested but should behave the same as platforms. Recommend at least extending the integration test to include a `[platform groups]` example to ensure there is some coverage of this feature.
* `[task events]` - From code inspection, this will probably get reloaded contrary to the documentation? Needs testing.

ScottWales commented:
For `[scheduler]`, it might be easier if `glbl_cfg(reload=True)` just doesn't update that section? That would avoid a mix of settings that are and aren't reloadable, e.g. right now `[scheduler][mail]task event batch interval` is cached by the `Scheduler` so not reloadable but `[scheduler][mail]from` is not. Although that's the case currently for the workflow configuration as well which these act as defaults for. I'll update the docs to mark `[scheduler][mail]` and `[scheduler][events]` as special cases.

Added a functional test for `reload --global` that shows `[platform]` and `[task events]` being updated. I don't think `[task events]` is documented as not being reloadable?

Added a integration test for platform group updates.

oliver-sanders commented on cylc/flow/scripts/reload.py:
```diff
@@ -60,19 +60,34 @@
 ) {
   reload (
     workflows: $wFlows
+    reloadGlobal: $reloadGlobal
```
[In relation to above comment]

I think we can avoid inter-version interoperability issues providing we omit the new `reloadGlobal` field when `--global` was not specified.

oliver-sanders commented:
@ScottWales We are starting to line up the major PRs for 8.5.0 release, which will hopefully take place over the next few weeks.

It would be good to get this into 8.5.0, otherwise it'll waiting for 8.6.0 which will probably be a few months further down the line.

ScottWales commented on cylc/flow/commands.py:
```diff
@@ -361,6 +362,25 @@ async def reload_workflow(schd: 'Scheduler'):
+                ' workflow.'
+                '\nOtherwise, fix the configuration and attempt to reload'
+                ' again.'
+            )
```
I've reworked this section to work as described, moving the global reload into the same `try` block as the workflow reload and rolling it back in the exception handler

ScottWales commented on cylc/flow/network/schema.py:
```diff
@@ -1919,6 +1919,10 @@ class Meta:
     class Arguments:
         workflows = graphene.List(WorkflowID, required=True)
 
+        reload_global = Boolean(
```
Added a fix to use reloadGlobal in the mutation only if `--global` was requested.

Tested reloading a new server with old cylc works fine. Reloading a old server with the new cylc works when `--global` is not used, when it is used you get a message
```
$ cylc reload --global reload-test
Error processing command:
    AttributeError: 'list' object has no attribute 'values'
```
and in the scheduler log
```
2025-04-02T11:20:41+11:00 ERROR - Unknown argument "reloadGlobal" on field "reload" of type "Mutations".
2025-04-02T11:20:41+11:00 ERROR - graphql.error.base.GraphQLError: Unknown argument "reloadGlobal" on field "reload" of type "Mutations".
```

ScottWales commented:
Thanks @oliver-sanders I've addressed the comments, sorry for the delay. If you're ok with the changes I'll rebase and fix the tests

oliver-sanders reviewed (approved):
Nice.

Tested:
* Workflow config reload.
* Global config rollback :+1:.
* Forward and backward compatibility.

LGTM.

A new lint error has cropped up on master so you might want to rebase again (sorry), the remainder are small lint errors to be fixed on this branch (we've turned on flake8 in a lax configuration for the integration tests now).

ScottWales commented:
Thanks, CI has been fixed and a doc update added at cylc/cylc-doc#815

hjoliver commented:
Brilliant, thanks @ScottWales

hjoliver commented on cylc/flow/commands.py:
```diff
@@ -360,12 +361,24 @@ async def reload_workflow(schd: 'Scheduler'):
+
+        if reload_global:
+            # Reload global config if requested
+            schd.reload_pending = 'reloading the global configuration'
```
The content and use of these `self.reload_pending` strings is a bit messy, but partly historical so I'll leave it be on this branch (and it's of no real consequence).

hjoliver reviewed (approved):
All good 👍
